src/kernels/kernelWorkingFolder.node.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as path from '../platform/vscode-path/path';
+import { inject, injectable } from 'inversify';
+import { Uri } from 'vscode';
+import { IWorkspaceService } from '../platform/common/application/types';
+import { getFilePath } from '../platform/common/platform/fs-paths';
+import { IFileSystem } from '../platform/common/platform/types';
+import { IConfigurationService, Resource } from '../platform/common/types';
+import { isLocalHostConnection, isPythonKernelConnection } from './helpers';
+import { expandWorkingDir } from './jupyter/jupyterUtils';
+import { IKernel, isLocalConnection, StartupCodePriority } from './types';
+// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
+const untildify = require('untildify');
+
+@injectable()
+export class KernelWorkingFolder {
+    public priority = StartupCodePriority.Base;
+
+    constructor(
+        @inject(IConfigurationService) private readonly configService: IConfigurationService,
+        @inject(IFileSystem) private readonly fs: IFileSystem,
+        @inject(IWorkspaceService) private readonly workspace: IWorkspaceService
+    ) {}
+
+    async getWorkingDirectory(kernel: IKernel): Promise<Uri | undefined> {
```

Comment:
Now we have sone single place to calculate the working directory and one way to access this via this method.